### ISSUE TITLE
Paper 390

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#122]
This issue is for discussion related to PR #6 

Author contact: @ForeverZyh

**Comment by kstolee** [comment]
## Authors

| Name                                 | email              | Github id    |
| ------------------------------------ | ------------------ | ------------ |
| Yuhao Zhang (*corresponding author*) | yuhaoz@cs.wisc.edu | ForeverZyh   |
| Luyao Ren                            | rly@pku.edu.cn     | luyaor       |
| Liqian Chen                          | lqchen@nudt.edu.cn | lqchen       |
| Yingfei Xiong                        | xiongyf@pku.edu.cn | xiongyingfei |
| Shing-Chi Cheung                     | scc@cse.ust.hk     | sccust       |
| Tao Xie                              | taoxie@gmail.com   | taoxiease    |

**Comment by kstolee** [comment]
Also PR #3

**Comment by art-reviewer** [comment]
[Paper](https://github.com/ForeverZyh/fse20/blob/e756d1ba8fd44ce30ca67f31c61abc4a921e8785/submissions/reusable/YuhaoZhang_fse20main-p390-p/fse20main-p390-p_submission.pdf)

**Comment by letmecheckthat** [comment]
Thank you for uploading this artifact. Upon first inspection, replicating the published results was successful and greatly facilitated by the use of Docker; I appreciate that. The instructions were clear and direct, and the code looks by and large well documented. One minor note: several files have out-commented blocks of logic (incl. `main.py`, `analysis_main.py`) that may be obsolete and could be cleaned up to improve the readability.

As for badges, this work is in a somewhat odd position: I would argue that it qualifies for both Functional and/or Reusable, but these were not requested. Is there any interest in these? On the other hand, Available was requested, but since the replication relies on an image from Dockerhub (to the best of my knowledge, not an archival repository), it does not yet qualify for that. Please let me know if I overlooked something. A simple fix would be to either upload the (pre-built) Docker image to e.g. Zenodo, or create a Zenodo archive that tracks the releases from the Github repo (and create a release). Please note that the datasets should also be stored on e.g. Zenodo, as Google Drive is not an archival repository either.

**Comment by FSEArtifactReviewer835** [comment]
Requested Badge: Reusable

Overall/Summary:  
The package is a tool to run several static analyses for numerical bugs in neural architectures.
The authors ran the tool on several datasets with known bugs along with many real-world projects.
The tool was contained in docker, and it was easy to run and replicate the results presented in the paper, including swapping over to other branches to look at variants (one small exception*, see the details). A skim through some of the code files seems to show that they are reasonably well-formatted and documented. In my opinion, this package meets the requirements for a reusable badge.


Basic Requirements:
CONTACT.md - Present, author name, emails, and GitHub id present
README.md - Present
INSTALL.md - Present
STATUS.md - Present, describes the requested badge (Reusable) and justification for why they should
receive it.
LICENSE.md - Present, Apache License 2.0
Accepted Paper - Present


Details:
The tool is provided in docker format, the docker file builds successfully.  However,
I would recommend posting a built docker to DockerHub as well.

The analysis_main.py example replicates as described in the install.md.

true_positives.csv - Contains manual classifications of true positives and false positives for the warnings.  The counts of true positives align with what is reported in the paper.

I was able to successfully run DEBAR (and the variants) and spot-checked some of the results for the first, second, and third section of Table 1.  

*I have also run the variant used to produce Table 3.  It appears to be running correctly but is taking a long time to complete so I haven't been able to verify the results in the same way.

**Comment by ForeverZyh** [comment]
Thanks for reviewing the artifact. 
To @letmecheckthat 's comments:

1. out-commented blocks of logic: we will delete these out-commented blocks of logic.
2. not requested Functional or Reusable: we also applied the reusable badge see PR #6. Apologize for creating the separate PR.
3. not yet qualify for the available badge: we uploaded the repository and dataset on Zenodo, please see https://github.com/researchart/fse20/blob/master/submissions/available/YuhaoZhang_fse20main-p390-p/index.md
We will upload the image to DockerHub as well, as both reviewers suggested.

**Comment by ForeverZyh** [comment]
To @FSEArtifactReviewer835 's comments:

1. posting a built docker to DockerHub: We will upload the image to DockerHub as well, as both reviewers suggested.
2. It takes a long time to produce Table 3 results. We found that we accidentally included 3 architectures that are supposed to time out for 30mins in the artifact. Now they've been removed, so please try again.

**Comment by ForeverZyh** [comment]
We've uploaded `debar` to DockerHub and you may access these images by

```
docker pull yuhaoz/debar:main
docker pull yuhaoz/debar:smashing-affine
docker pull yuhaoz/debar:partition-wo-affine
docker pull yuhaoz/debar:expansion-affine
```

Also, if you want to initiate a bash into the image you may change the previous `lastest` tag by the new tag, e.g.,

```
docker run -it yuhaoz/debar:main bash
```

**Comment by letmecheckthat** [comment]
Great, looks like Available and Reusable are appropriate as far as I'm concerned. Thanks for the updates!

**Comment by FSEArtifactReviewer835** [comment]
I agree, I think both Available and Reusable are appropriate badges.

**Comment by kstolee** [comment]
Thanks @letmecheckthat and @FSEArtifactReviewer835 for the reviews! And to @ForeverZyh for being so responsive.
